src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -1684,6 +1822,132 @@ UndistributeTable(Oid relationId)
+ */
+void
+UndistributeTable(Oid relationId)
+{
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("cannot undistribute table"
+							   "because no such distributed table exists")));
+	}
+
+	relation_close(relation, NoLock);
+
+	if (!IsCitusTable(relationId))
+	{
+		ereport(ERROR, (errmsg("cannot undistribute table "
+							   "because the table is not distributed")));
+	}
+
+	EnsureTableNotReferencing(relationId);
+	EnsureTableNotReferenced(relationId);
+	EnsureTableNotForeign(relationId);
+	EnsureTableNotPartition(relationId);
+
+	ConvertTable(UNDISTRIBUTE_TABLE, relationId, NULL, 0, NULL, NULL, false);
+}
+
+
+/*
+ * AlterDistributedTable changes some properties of the given table. It uses
+ * ConvertTable function to create a new local table and move everything to that table.
+ *
+ * The local and reference tables, tables with references, partition tables and foreign
+ * tables are not supported. The function gives errors in these cases.
+ */
+void
+AlterDistributedTable(Oid relationId, char *distributionColumn, int shardCount,
+					  char *colocateWithTableName, bool cascadeToColocated)
+{
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("cannot undistribute table "
+							   "because no such distributed table exists")));
+	}
+	relation_close(relation, NoLock);
+
+	if (!IsCitusTableType(relationId, DISTRIBUTED_TABLE))
+	{
+		ereport(ERROR, (errmsg("cannot undistribute table "
+							   "because the table is not distributed")));
+	}
+
+	EnsureTableNotReferencing(relationId);
+	EnsureTableNotReferenced(relationId);
+	EnsureTableNotForeign(relationId);
+	EnsureTableNotPartition(relationId);
+
+	ConvertTable(ALTER_DISTRIBUTED_TABLE, relationId, distributionColumn, shardCount,
+				 colocateWithTableName, NULL, cascadeToColocated);
+}
+
+
+/*
+ * AlterTableSetAccessMethod changes the access method of the given table. It uses
+ * ConvertTable function to create a new table with the access method and move everything
+ * to that table.
+ *
+ * The local and references tables, tables with references, partition tables and foreign
+ * tables are not supported. The function gives errors in these cases.
+ */
+void
+AlterTableSetAccessMethod(Oid relationId, char *accessMethod)
+{
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("cannot undistribute table"
+							   "because no such distributed table exists")));
+	}
+	relation_close(relation, NoLock);
+
+	if (!IsCitusTableType(relationId, DISTRIBUTED_TABLE))
+	{
+		ereport(ERROR, (errmsg("cannot undistribute table "
+							   "because the table is not distributed")));
+	}
+
+	EnsureTableNotReferencing(relationId);
+	EnsureTableNotReferenced(relationId);
+	EnsureTableNotForeign(relationId);
+	EnsureTableNotPartition(relationId);
```

Comment:
ideally, we would also support this for partitions by doing the attach/detach in:
https://gist.github.com/marcocitus/6b0fb4ffbb38d52456283162b2679506#file-alter_table_access_method-sql-L41

The typical way of using columnar storage will be as part of a partitioned table in which older partitions are converted using `alter_table_set_access_method`.